File: src/main/java/es/danisales/datune/tempo/BPM.java
package es.danisales.datune.tempo;

import java.util.Map;


public class BPM extends Tempo {
	private int bpm;
	private MusicalTime base;

	private BPM() { }

	public static BPM of(int q) {
		return of(q, MusicalTime.QUARTER);
	}

	@SuppressWarnings("WeakerAccess")
	public static BPM of(int q, MusicalTime sd) {
		BPM bpm = new BPM();
		bpm.bpm = q;
		bpm.base = sd;

		bpm.initialize();
		return bpm;
	}

	@Override
	public void init(Map<Double, MusicalTime> map) {
		final double msPerMin = 1000 * 60;
		final double baseDuration = msPerMin/bpm;
		final double wholeDuration = baseDuration / base.val;
		map.put( baseDuration, base );

		for (MusicalTime d : MusicalTime.values())
			if (d.equals(base))
				map.put( wholeDuration * d.val, d);
	}
}

File: src/tempo/BPM.ts
import { MusicalDuration } from "./MusicalDuration";
import { Tempo } from "./Tempo";

export class BPM extends Tempo {
    public static QUARTER_120: BPM;

    private constructor(private _bpm: number, private _beat: MusicalDuration) {
        super();
    }

    public static from(bpm: number, beat: MusicalDuration = MusicalDuration.QUARTER): BPM {
        return new BPM(bpm, beat);
    }

    public getMillis(musicalDuration: MusicalDuration): number {
        let msPerMin = 1000 * 60;
        let baseDuration = msPerMin / this._bpm;
        let wholeDuration = baseDuration / this._beat.value;

        return musicalDuration.value * wholeDuration;
    }

    get bpm(): number {
        return this._bpm;
    }

    get beat(): MusicalDuration {
        return this._beat;
    }

    private static initialize() {
        this.QUARTER_120 = BPM.from(120, MusicalDuration.QUARTER);
    }
}